### ISSUE TITLE
LabBuilder.py: The domain name label var.prefix-XXX is invalid

### ISSUE THREAD

**Issue by tsluyter** [DefensiveOrigins/APT-Lab-Terraform#7]
When running LabBuilder.py the initial setup works correctly, but the first step in creating the VMs fails. When defining the network interfaces, I am given the following error messages:

```
Error: network.InterfacesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidDomainNameLabel" Message="The domain name label var.prefix-dc is invalid. It must conform to the following regular expression: ^[A-Za-z0-9][A-Za-z0-9\\-]{1,62}(?<!\\-)$." Details=[]

  on modules/active-directory/1-network-interface.tf line 1, in resource "azurerm_network_interface" "primary":
   1: resource "azurerm_network_interface" "primary" {

Error: network.InterfacesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidDomainNameLabel" Message="The domain name label var.prefix-linux is invalid. It must conform to the following regular expression: ^[A-Za-z0-9][A-Za-z0-9\\-]{1,62}(?<!\\-)$." Details=[]

  on modules/linux/1-network-interface.tf line 1, in resource "azurerm_network_interface" "primary":
   1: resource "azurerm_network_interface" "primary" {

Error: network.InterfacesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidDomainNameLabel" Message="The domain name label var.prefix-client is invalid. It must conform to the following regular expression: ^[A-Za-z0-9][A-Za-z0-9\\-]{1,62}(?<!\\-)$." Details=[]

  on modules/windows-client1/1-network-interface.tf line 8, in resource "azurerm_network_interface" "primary":
   8: resource "azurerm_network_interface" "primary" {
```

I've taken a look at the relevant configuration files (for example _LABS/modules/active-directory/1-network-interface.tf_) where indeed `"var.prefix-dc"` is used for the name field. I wonder whether it's not resolving the variable, but instead is inserting the variable name as literal string.

**Comment by tsluyter** [comment]
Resolved the issue by manually tweaking the `1-network*` files, e.g. `master/modules/*/1-network*` as well as the `2-virtual-machine*` files. In both cases the "name" field pointed at a non-existent variable and the string was being used literally. For some reason Terraform and Azure weren't having fun defining a hostname of "var.prefix-dc.local".

**Comment by tsluyter** [comment]
Yeah I see what happened -> https://github.com/DefensiveOrigins/APT-Lab-Terraform/commit/b58a0cd6bf2f36f177a695b1244e7f840da857a8#diff-6c6baf5519de246eb87126c8d2de91dc41fb567e39e8b3aedf64fea984c464db

In that commit, the `${ ... }` were removed from all variable names. I don't know whether that plays nicely with Terraform on your system, but it does not on MacOS with the TF installed through Brew.

The same mishap also affects things like storage provisioning, where the same name gets re-used because the variable isn't being expanded. For example from the provisioning of the AD DC VM:

```
Provisioning failed.
A disk with name local.virtual_machine_name-disk1 already exists in Resource Group CLASS-RESOURCES and is attached to VM /subscriptions/aaa-bbb-ccc-ddd-eee-fffffffff/resourceGroups/class-resources/providers/Microsoft.Compute/virtualMachines/stu-linux. 'Name' is an optional property for a disk and a unique name will be generated if not provided.  Target: '/subscriptions/aaa-bbb-ccc-ddd-eee-fffffffff/resourceGroups/class-resources/providers/Microsoft.Compute/disks/local.virtual_machine_name-disk1'..
ConflictingUserInput
```

Issues occur whenever a `local.xxx` or `var.xxx` is called within a double-quoted string, without being enclosed in `${ ... }`. I'm trying to edit the relevant .tf files to implement fixes locally; see if I can get it to run again.

**Comment by tsluyter** [comment]
Can be closed when pull req #9 is applied.

**Comment by tsluyter** [comment]
Tackled by completed merge.
